### ISSUE TITLE
Segmentation fault (Spacer)

### ISSUE THREAD

**Issue by leonardoalt** [Z3Prover/z3#3526]
Query: https://gist.github.com/leonardoalt/bd5686468e4831a68e82afdeacb98c95
z3 version: Z3 version 4.8.7 - 64 bit
Command line: z3 fp.engine=spacer fp.spacer.q3.use_qgen=true fp.spacer.mbqi=false fp.spacer.ground_pobs=false evm_chc.smt2
OS: Arch Linux

On my machine it seg faults after about 3s. I haven't tried to simplify it yet, but removing the parameters make it hang longer, not sure if it crashes after.

**Comment by agurfinkel** [comment]
Hey @leonardoalt , you are using spacer with datatypes and strings. This is not supported. 
I'll see if I can fix the segfault. But the fix will most likely either cause spacer to return `unknown` or hang.

**Comment by leonardoalt** [comment]
Ah ok, thanks!
Are datatypes without strings supported?

**Comment by agurfinkel** [comment]
Do not know. You would be the only user.

**Comment by NikolajBjorner** [comment]
it isn't crashing on my side (latest master).
ADTs lacks tuning (inductive generalization is super weak), 
and testing on larger suite of benchmarks.

**Comment by leonardoalt** [comment]
Thanks @NikolajBjorner 
I was running on release, but I'll keep a lib/binary from `master` as well

**Comment by NikolajBjorner** [comment]
I am now seeing a crash in debug mode after a while.
It is either stack abuse or infinite loop (causing a stack overflow).

**Comment by agurfinkel** [comment]
`qe_term_graph` gets into an infinite loop because it picks a wrong representative. 
I was debugging this until I ran out of time. Here is the graph term table
```
157: state_0_n R -
1081: (pc state_0_n) - 3057
45: 2 R -
8693: (+ 2 (pc state_0_n)) R -
1082: (code state_0_n) R -
8694: (str.at (code state_0_n) (+ 2 (pc state_0_n))) R -
8695: (str.to.int (str.at (code state_0_n) (+ 2 (pc state_0_n)))) - 18729 33
33: 1 R - 8695 18729
3053: (seq.pre (code state_0_n) (pc state_0_n)) R -
3057: (str.len (seq.pre (code state_0_n) (pc state_0_n))) R - 1081
1262: 4 R -
18726: (+ 4 (pc state_0_n)) R -
18728: (str.at (code state_0_n) (+ 4 (pc state_0_n))) R -
18729: (str.to.int (str.at (code state_0_n) (+ 4 (pc state_0_n)))) - 33 8695
1731: (str.len (code state_0_n)) R -
19358: (<= (str.len (code state_0_n)) (pc state_0_n)) R -
19709: (not (<= (str.len (code state_0_n)) (pc state_0_n))) R -
```

**Comment by agurfinkel** [comment]
`1081` and `3057` are in the same equivalence class. `1081` should be the representative because it is smaller. But it is not.

**Comment by NikolajBjorner** [comment]
@leonardoalt - best to avoid strings when applying the SPACER engine.
There is no support for invariant synthesis with strings and reasoning with strings is relatively high overhead. It is likely more useful to create suitable abstractions from the caller side. More work for the user for sure, but this is frankly the first time I see someone throw strings at the Horn engine.

**Comment by leonardoalt** [comment]
@NikolajBjorner no worries! I'll switch to something more reasonable, thanks!
(Please feel free to close the issue)

**Comment by NikolajBjorner** [comment]
we should address the infinite loop (stack overflow) before moving on.

**Comment by agurfinkel** [comment]
once character typo. Committing a fix

**Comment by agurfinkel** [comment]
fixed by 5673ec046bde06cfd4aa2e70d1318fe40c84e4d4
